yarnpkg/yarn
# Fix "Cannot read property 'split' of undefined" for interconnected file based dependencies

**Summary**
Currently, when you have interconnected file dependencies, yarn install fails after a lockfile has been generated. This failure occurs because yarn calls `split` on the package's resolved url, which is `undefined`. This fix makes sure we have a resolved url before trying to call `split` on it.



This is described in #2256. You can find a repo that reproduces this issue at https://github.com/mstijak/yarn-file-dep-repro

As described in #2256, running `yarn install` on the above repo produces the following lockfile:

```
# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
# yarn lockfile v1


"a@file:./packages/a":
  version "1.0.2"
  dependencies:
    b "1.0.0"

b@*:
  version "1.0.0"

"b@file:./packages/b":
  version "1.0.0"
```

When you run `yarn install` a second, third, fourth, etc. time, yarn fails when trying to resolve the following package:

```
b@*:
  version "1.0.0"
```

It attempts to use a tarball resolver. In doing so, it calls split on a url that is undefined. My fix just checks makes sure the url is defined before trying to explode it. I'm not terribly familiar with yarn's code, so I'm not sure if this is a good fix or not.

**Test plan**
*To reproduce the bug*
1. Check out https://github.com/mstijak/yarn-file-dep-repro
2. Run `yarn install`. Notice that `yarn.lock` now exists.
3. Run `yarn install` again. This should fail with the following error
```
Trace: 
  TypeError: Cannot read property 'split' of undefined
      at Object.explodeHashedUrl (/home/james/yarn/lib/util/version.js:8:20)
      at PackageRequest.getLocked (/home/james/yarn/lib/package-request.js:128:59)
      at /home/james/yarn/lib/resolvers/registries/npm-resolver.js:202:37
      at next (native)
      at step (/home/james/yarn/node_modules/babel-runtime/helpers/asyncToGenerator.js:17:30)
      at /home/james/yarn/node_modules/babel-runtime/helpers/asyncToGenerator.js:35:14
      at Promise.F (/home/james/yarn/node_modules/core-js/library/modules/_export.js:35:28)
      at /home/james/yarn/node_modules/babel-runtime/helpers/asyncToGenerator.js:14:12
      at NpmResolver.resolve (/home/james/yarn/lib/resolvers/registries/npm-resolver.js:237:7)
      at /home/james/yarn/lib/package-request.js:183:23
```
*To verify the fix*
With the fix applied repeat the steps above. On step 3, `yarn install` should succeed without error.

**Gifs demonstrating the fix follow**
*Broken*
![broken](http://i.imgur.com/eNVri94.gif)

*Fixed*
![fixed](http://i.imgur.com/qPjzamk.gif)
